File: source/src/lib/api/science.ts
import { invoke } from "@tauri-apps/api/core";

export interface ScienceStatus {
  supported: boolean;
  installed: boolean;
  running: boolean;
  healthy: boolean;
  port: number;
  providerName?: string;
  runtimeSource?: string;
  runtimeVersion?: string;
  message?: string;
}

export interface ScienceStartResult {
  url: string;
  providerName: string;
  runtimeSource: string;
  runtimeVersion: string;
}

export const scienceApi = {
  getStatus(): Promise<ScienceStatus> {
    return invoke("get_science_status");
  },
  start(): Promise<ScienceStartResult> {
    return invoke("start_science");
  },
  stop(): Promise<void> {
    return invoke("stop_science");
  },
  open(): Promise<string> {
    return invoke("open_science");
  },
};

File: source/src-tauri/src/commands/science.rs
use crate::science::{self, ScienceStartResult, ScienceStatus};
use crate::store::AppState;

#[tauri::command]
pub async fn get_science_status(
    state: tauri::State<'_, AppState>,
) -> Result<ScienceStatus, String> {
    Ok(science::status(&state).await)
}

#[tauri::command]
pub async fn start_science(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
) -> Result<ScienceStartResult, String> {
    science::start(&app, &state).await
}

#[tauri::command]
pub async fn stop_science() -> Result<(), String> {
    science::stop().await
}

#[tauri::command]
pub async fn open_science(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
) -> Result<String, String> {
    science::open(&app, &state).await
}
